File: jscomp/test/imm_map_bench.js
'use strict';

var Immutable = require("immutable");
var Belt_Array = require("melange/lib/js/belt_Array.js");
var Belt_MapInt = require("melange/lib/js/belt_MapInt.js");

var empty = new Immutable.OrderedMap();

function fromArray(kvs) {
  var v = empty;
  for(var i = 0 ,i_finish = kvs.length; i < i_finish; ++i){
    var match = kvs[i];
    v = v.set(match[0], match[1]);
  }
  return v;
}

function should(b) {
  if (b) {
    return ;
  }
  throw new Error("impossible");
}

var shuffledDataAdd = Belt_Array.makeByAndShuffle(1000001, (function (i) {
        return [
                i,
                i
              ];
      }));

function test(param) {
  var v = fromArray(shuffledDataAdd);
  for(var j = 0; j <= 1000000; ++j){
    should(v.has(j));
  }
}

function test2(param) {
  var v = Belt_MapInt.fromArray(shuffledDataAdd);
  for(var j = 0; j <= 1000000; ++j){
    should(Belt_MapInt.has(v, j));
  }
}

console.time("imm_map_bench.ml 44");

test(undefined);

console.timeEnd("imm_map_bench.ml 44");

console.time("imm_map_bench.ml 45");

test2(undefined);

console.timeEnd("imm_map_bench.ml 45");

var count = 1000000;

exports.empty = empty;
exports.fromArray = fromArray;
exports.should = should;
exports.count = count;
exports.shuffledDataAdd = shuffledDataAdd;
exports.test = test;
exports.test2 = test2;
/* empty Not a pure module */

File: jscomp/test/imm_map_bench.ml

type map
external empty : unit -> map = "OrderedMap" [@@mel.new] [@@mel.module "immutable"]
external set : map -> int -> int -> map = "set" [@@mel.send]
external get : map -> int -> int option = "get" [@@mel.send] [@@mel.return undefined_to_opt]
external mem : map -> int -> bool = "has" [@@mel.send]


module A = Belt.Array
let empty = empty ()
let fromArray kvs =
  let v = ref empty in
  for i = 0 to A.length kvs - 1 do
    let key, value = (A.getUnsafe kvs i)  in
    v := set !v key value
  done;
  !v

let should b =
  if not b  then Js.Exn.raiseError "impossible"

let count = 1_000_000

let shuffledDataAdd = (A.makeByAndShuffle (count +  1) (fun i -> (i,i)))



let test () =
  let v = fromArray shuffledDataAdd  in
  for j = 0 to count do
    should (mem v j)
  done

module M =  Belt.Map.Int


let test2 () =
  let v = M.fromArray shuffledDataAdd in
  for j = 0 to count do
    should (M.has v j)
  done ;


;; [%time test ()]
;; [%time test2 ()]
